### PR TITLE
Feature/add the media manager

### PR DESCRIPTION
**Pull request by dchvs99**

### CODE REVIEW

File: tests/gstreamer/test_mediamanager.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+import gi  # nopep8
+gi.require_version('Gst', '1.0')  # nopep8
+gi.require_version('GLib', '2.0')  # nopep8
+from gi.repository import Gst as gst  # nopep8
+from gi.repository import GLib  # nopep8
+
+import random
+from rr.gstreamer.mediamanager import MediaManager
+from rr.gstreamer.mediamanager import MediaManagerError
+import unittest
```
Comment (emmadrigal) [inline tests/gstreamer/test_mediamanager.py]: What include order are you following?

Python standard would be:
```
Python standard libs

external libs

internal includes

```

File: rr/gstreamer/mediamanager.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+
+class MediaManager():
+    """
+    Class that handles the medias
+
+    Attributes
+    ----------
+    _pipeline : GstElement
+        A private GStreamer pipeline object
```
Comment (emmadrigal) [inline rr/gstreamer/mediamanager.py]: `_pipeline`?

you are missing `_Dict`

File: tests/gstreamer/test_mediamanager.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        # Check dictionary it's empty
+        dict = self.mediamanager._get_media_dict()
+        self.assertEqual(0, len(dict))
+
+    def testremove_media(self):
+        with self.assertRaisesRegex(MediaManagerError, "Invalid key"):
+            self.mediamanager.remove_media(None)
+
+        with self.assertRaisesRegex(MediaManagerError, "Unable to find the key in the dictionary"):
+            self.mediamanager.remove_media(random.random())
```
Comment (emmadrigal) [inline tests/gstreamer/test_mediamanager.py]: Why random? 

You have an empty dictionary. Or you have inserted a media with a known key?
In both cases you can use a key known by you, for example `-1`
But in the second the random number can match your known key.

File: tests/gstreamer/test_mediamanager.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+    def teststart_media(self):
+        key = "start_media"
+        with self.assertRaisesRegex(MediaManagerError, "Invalid key or media"):
+            self.mediamanager.add_media(key, None)
+
+            with self.assertRaisesRegex(MediaManagerError, "Unable to start media"):
```
Comment (emmadrigal) [inline tests/gstreamer/test_mediamanager.py]: Is this assertion being tested correctly? If the first one fails (which appears to be expected), does the flow even reach this call?

**Review by emmadrigal — CHANGES_REQUESTED** [review]

File: tests/gstreamer/test_mediamanager.py (hunk 3 above)
Comment (dchvs99) [reply tests/gstreamer/test_mediamanager.py]: It is an empty dictionary, and having a random nonexistent key just adds up some entropy.

File: rr/gstreamer/mediamanager.py [hunk 5]
```diff
@@ -0,0 +1,135 @@
+#!/usr/bin/env python3
```
Comment (michaelgruner) [inline rr/gstreamer/mediamanager.py]: Please discuss with @Solzc what module naming convention we are going to adopt. I believe @Solzc is using `snake_case`.

File: rr/gstreamer/mediamanager.py [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+        MediaManagerError
+            If the description fails to create the media
+
+        Return
+        ------
+        media : GstMedia obj from a description
+        """
+
+        try:
+            self.gst_media = GstMedia()
```
Comment (michaelgruner) [inline rr/gstreamer/mediamanager.py]: Nope, the media is created outside. This way we can install different medias.

File: rr/gstreamer/mediamanager.py [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+        if key not in self._Dict:
+            raise MediaManagerError("Unable to find the key in the dictionary")
+
+        if self._Dict[key] is not None:
+            del self._Dict[key]
+            self._Dict[key] = None
+
+        self._Dict.pop(key)
+
+    def start_media(self):
```
Comment (michaelgruner) [inline rr/gstreamer/mediamanager.py]: medias use the `play` method. Why change it to `start`?

File: rr/gstreamer/mediamanager.py [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+
+        if key is None:
+            raise MediaManagerError("Invalid key")
+
+        if key not in self._Dict:
+            raise MediaManagerError("Unable to find the key in the dictionary")
+
+        if self._Dict[key] is not None:
+            del self._Dict[key]
+            self._Dict[key] = None
```
Comment (michaelgruner) [inline rr/gstreamer/mediamanager.py]: You don't want to delete it explicitly. Just set it to null. If you call `del` you remove it even if someone else is using it.

File: rr/gstreamer/mediamanager.py (hunk 7 above)
Comment (michaelgruner) [inline rr/gstreamer/mediamanager.py]: Please implement stop as well.

File: tests/gstreamer/test_gstmedia.py [hunk 9]
```diff
@@ -21,7 +22,7 @@ def _get_mediaState(media):
 
 class TestGstMedia(unittest.TestCase):
     def setUp(self):
-        self.desc = "videotestsrc ! fakesink"
+        self.desc = "videotestsrc ! fakesink async=false"
```
Comment (michaelgruner) [inline tests/gstreamer/test_gstmedia.py]: Why?

Comment (michaelgruner) [reply tests/gstreamer/test_gstmedia.py]: Also, use `is-live=true` in the videotestsrc.

**Review by michaelgruner — CHANGES_REQUESTED** [review]

File: rr/gstreamer/media_manager.py [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+        media : obj
+            The media object to add to dictionary
+
+        Raises
+        ------
+        MediaManagerError
+            If the description fails to insert the media
+        """
+
+        if key is None or media is None:
```
Comment (emmadrigal) [inline rr/gstreamer/media_manager.py]: surround the separate logic by parenthesis:
`if (key is None) or (media is None)`:

File: rr/gstreamer/imedia.py [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+        Raises
+        ------
+        MediaError
+            If the description fails to create the media
+        """
+
+        try:
+            self.media.create_media(desc)
+        except MediaError as e:
+            raise IMediaError("Unable to create the media")
```
Comment (emmadrigal) [inline rr/gstreamer/imedia.py]: When chaining exceptions (calling an exception from the catch of another exception) use `from` as follows
```
    raise IMediaError("Unable...") from e
```

File: rr/gstreamer/imedia.py [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+        Raises
+        ------
+        MediaError
+            If could not delete the media
+        """
+
+        try:
+            self.media.delete_media()
+        except MediaError as e:
+            raise IMediaError("Unable to delete the media")
```
Comment (emmadrigal) [inline rr/gstreamer/imedia.py]: `raise IMediaError() from e`

File: rr/gstreamer/imedia.py [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+        Raises
+        ------
+        MediaError
+            If couldn't set the media state to playing
+        """
+
+        try:
+            self.media.play_media()
+        except MediaError as e:
+            raise IMediaError("Unable to play the media")
```
Comment (emmadrigal) [inline rr/gstreamer/imedia.py]: `raise IMediaError() from e`

File: rr/gstreamer/imedia.py [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+
+        Raises
+        ------
+        MediaError
+            If couldn't set the media state to stopped
+        """
+        try:
+            self.media.stop_media()
+        except MediaError as e:
+            raise IMediaError("Unable to stop the media")
```
Comment (emmadrigal) [inline rr/gstreamer/imedia.py]: `raise IMediaError() from e`

File: rr/gstreamer/imedia.py [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+
+    def get_media(self):
+        """Getter for the media object
+        """
+
+        try:
+            ret = self.media.get_media()
+        except MediaError as e:
+            raise IMediaError("Invalid media")
```
Comment (emmadrigal) [inline rr/gstreamer/imedia.py]: `raise IMediaError() from e`

File: rr/gstreamer/media_manager.py [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+        ------
+        MediaManagerError
+            If the description fails to play the medias
+        """
+
+        for key in self._Dict:
+            try:
+                self._Dict[key].play_media()
+            except IMediaError as e:
+                raise MediaManagerError("Unable to start media") from e
```
Comment (emmadrigal) [inline rr/gstreamer/media_manager.py:103]: `raise MediaManagerError() from e`

File: rr/gstreamer/media_manager.py [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+        ------
+        MediaManagerError
+            If the description fails to stop the medias
+        """
+
+        for key in self._Dict:
+            try:
+                self._Dict[key].stop_media()
+            except IMediaError as e:
+                raise MediaManagerError("Unable to stop media") from e
```
Comment (emmadrigal) [inline rr/gstreamer/media_manager.py:118]: `raise MediaManagerError() from e`

File: rr/gstreamer/imedia.py [hunk 18]
```diff
@@ -0,0 +1,119 @@
...
+
+#  Copyright (C) 2021 RidgeRun, LLC (http://www.ridgerun.com)
+#  Authors: Daniel Chaves <daniel.chaves@ridgerun.com>
+#           Marisol Zeledon <marisol.zeledon@ridgerun.com>
+
+import gi  # nopep8
+gi.require_version('Gst', '1.0')  # nopep8
+gi.require_version('GLib', '2.0')  # nopep8
+from gi.repository import Gst as gst  # nopep8
+from gi.repository import GLib  # nopep8
```
Comment (emmadrigal) [inline rr/gstreamer/imedia.py]: Are all of these being used?

File: rr/gstreamer/media_manager.py [hunk 19]
```diff
@@ -0,0 +1,127 @@
...
+
+#  Copyright (C) 2021 RidgeRun, LLC (http://www.ridgerun.com)
+#  Authors: Daniel Chaves <daniel.chaves@ridgerun.com>
+#           Marisol Zeledon <marisol.zeledon@ridgerun.com>
+
+import gi  # nopep8
+gi.require_version('Gst', '1.0')  # nopep8
+gi.require_version('GLib', '2.0')  # nopep8
+from gi.repository import Gst as gst  # nopep8
+from gi.repository import GLib  # nopep8
```
Comment (emmadrigal) [inline rr/gstreamer/media_manager.py]: Are all of these being used?

File: rr/gstreamer/imedia.py [hunk 20]
```diff
@@ -0,0 +1,119 @@
...
+        MediaError
+            If couldn't set the media state to stopped
+        """
+        try:
+            self.media.stop_media()
+        except MediaError as e:
+            raise IMediaError("Unable to stop the media")
+
+    def get_media(self):
+        """Getter for the media object
```
Comment (emmadrigal) [inline rr/gstreamer/imedia.py:105]: Missing `Returns` field in docstring

File: tests/gstreamer/test_media_manager.py [hunk 21]
```diff
@@ -0,0 +1,113 @@
...
+
+#  Copyright (C) 2021 RidgeRun, LLC (http://www.ridgerun.com)
+#  Authors: Daniel Chaves <daniel.chaves@ridgerun.com>
+#           Marisol Zeledon <marisol.zeledon@ridgerun.com>
+
+import gi  # nopep8
+gi.require_version('Gst', '1.0')  # nopep8
+gi.require_version('GLib', '2.0')  # nopep8
+from gi.repository import Gst as gst  # nopep8
+from gi.repository import GLib  # nopep8
```
Comment (emmadrigal) [inline tests/gstreamer/test_media_manager.py]: Are all of these being used?

File: tests/gstreamer/test_media_manager.py [hunk 22]
```diff
@@ -0,0 +1,113 @@
...
+
+import random
+import unittest
```
Comment (emmadrigal) [inline tests/gstreamer/test_media_manager.py:8]: python's imports should be at the top (before any other library)

**Review by emmadrigal — CHANGES_REQUESTED** [review]

**Pull request merged**